File: packages/vue-im-ui/src/utils/imageGallery.ts
import type { FlareMediaResolveRequest } from "../shared/contracts/media";
import { resolveMessageId, type MessageLike } from "../shared/contracts/messageRow";
import { asRecord, readArray, readString } from "./contentData";
import { normalizeToContentElem, pickNestedPayload } from "./contentElem";
import {
  imageGroupItemRequest,
  imageGroupItemThumbnailRequest,
  imageMessagePayload,
  imageSourceRequest,
  imageThumbnailRequest,
} from "./imageRequests";

/**
 * A conversation's image gallery: the pictures a full-screen preview opened from a timeline pages through. The rule
 * is shared with the three native kits (`spec/image-gallery-vectors.json`).
 */
export interface FlareImageGalleryItem {
  /** The message (`resolveMessageId`) the picture belongs to. */
  messageId: string;
  /** The picture's place in its own message. */
  index: number;
  /** An image message's picture, or one picture of an album. */
  kind: "image" | "imageGroupItem";
  /** The picture's payload as the message carries it. */
  image: Record<string, unknown>;
}

/**
 * Every picture of every image and album message in `messages` (timeline order, oldest first), skipping recalled
 * messages and pictures with nothing to load — no address and no stored media id.
 */
export function flareImageGalleryItems(messages: readonly MessageLike[]): FlareImageGalleryItem[] {
  const items: FlareImageGalleryItem[] = [];
  for (const message of messages) {
    if (message.isRecalled) continue;
    const content = normalizeToContentElem(message.content);
    if (content?.contentType === "image") {
      const image = imageMessagePayload(content);
      const item: FlareImageGalleryItem = { messageId: resolveMessageId(message), index: 0, kind: "image", image };
      if (flareImageGalleryRequest(item)) items.push(item);
    } else if (content?.contentType === "image_group") {
      const nested = pickNestedPayload(content, "image_group");
      const album = Object.keys(nested).length ? nested : (content as Record<string, unknown>);
      readArray(album, "images").forEach((raw, index) => {
        const item: FlareImageGalleryItem = { messageId: resolveMessageId(message), index, kind: "imageGroupItem", image: asRecord(raw) };
        if (flareImageGalleryRequest(item)) items.push(item);
      });
    }
  }
  return items;
}

/** Where a gallery of `items` starts for a tap on the picture at `index` of message `messageId`; null when it is not in it. */
export function flareImageGalleryStart(items: readonly FlareImageGalleryItem[], messageId: string, index: number): number | null {
  const at = items.findIndex((item) => item.messageId === messageId && item.index === index);
  return at < 0 ? null : at;
}

/** What the preview loads for `item`: its full-size picture, else its thumbnail; null when there is nothing to load. */
export function flareImageGalleryRequest(item: FlareImageGalleryItem): FlareMediaResolveRequest | null {
  return item.kind === "image"
    ? imageSourceRequest(item.image, item.messageId) ?? imageThumbnailRequest(item.image, item.messageId)
    : imageGroupItemRequest(item.image, item.index, item.messageId) ??
        imageGroupItemThumbnailRequest(item.image, item.index, item.messageId);
}

/** The picture's description, as its body names it. */
export function flareImageGalleryAlt(item: FlareImageGalleryItem): string {
  return item.kind === "image"
    ? readString(item.image, "description", "caption", "title")
    : readString(item.image, "description", "title");
}

File: packages/ios-im-ui/Sources/FlareIMUI/Models/ImageGallery.swift
import Foundation

// A conversation's image gallery: the pictures a full-screen preview opened from a timeline pages through. The rule
// is shared with the other three kits (`spec/image-gallery-vectors.json`).

/// One picture of the gallery: the message it belongs to, its place in that message, and the picture.
public struct FlareImageGalleryItem: Equatable {
    public let messageId: String
    public let index: Int
    public let image: FlareImageContent

    public init(messageId: String, index: Int, image: FlareImageContent) {
        self.messageId = messageId; self.index = index; self.image = image
    }

    /// The address the preview loads: the full-size image, else its thumbnail.
    public var source: String {
        image.url.trimmingCharacters(in: .whitespaces).isEmpty
            ? (image.thumbnailURL ?? "").trimmingCharacters(in: .whitespaces) : image.url
    }

    public static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.messageId == rhs.messageId && lhs.index == rhs.index && lhs.source == rhs.source && lhs.image.alt == rhs.image.alt
    }
}

/// Every picture of every image and album message in `messages` (timeline order, oldest first), skipping recalled
/// messages and pictures with nothing to load.
public func flareImageGalleryItems(_ messages: [FlareMessageData]) -> [FlareImageGalleryItem] {
    var items: [FlareImageGalleryItem] = []
    for message in messages where !message.isRecalled {
        let images: [FlareImageContent]
        switch message.content {
        case let image as FlareImageContent: images = [image]
        case let album as FlareImageGroupContent: images = album.images
        default: images = []
        }
        for (index, image) in images.enumerated() {
            let item = FlareImageGalleryItem(messageId: message.id, index: index, image: image)
            if !item.source.isEmpty { items.append(item) }
        }
    }
    return items
}

/// Where a gallery of `items` starts for a tap on the picture at `index` of message `messageId`; nil when it is not
/// in the gallery.
public func flareImageGalleryStart(_ items: [FlareImageGalleryItem], messageId: String, index: Int) -> Int? {
    items.firstIndex { $0.messageId == messageId && $0.index == index }
}

/// The gallery of the timeline a message body is drawn in: the timeline's messages, read only when a picture opens,
/// and the host's download handler for a picture of one of them.
struct FlareImageGallerySource {
    let messages: [FlareMessageData]
    let download: ((FlareMessageData, FlareMessageContent) -> Void)?
}
